### ISSUE TITLE
Rust code generation issue with wit-bindgen-cli

### ISSUE THREAD

**Issue by ouillie** [bytecodealliance/wit-bindgen#1051]
I'm using the `wit-bindgen` CLI to generate rust bindings, as opposed to the crate and macros, and running into some issues with the generated code.

Steps to reproduce:

1. Use the `wit-bindgen` binary built at commit b09f234.
2. Take this example WIT file:

```wit
# foo.wit
package test:foo;

world the-world {
    export words: func(text: string) -> list<string>;
    import log-activity: func(text: string);
}
```

3. Run `wit-bindgen rust foo.wit --world the-world --out-dir . --generate-all --pub-export-macro --default-bindings-module the_world` to generate the file `the_world.rs`.
4. Try to compile `rustc the_world.rs --crate-type=rlib`. This will result in several errors -- mostly failing to resolve things like `::core` in the generated Rust code -- but the only one that matters for the purpose of this bug is this\*:

```
error[E0255]: the name `__export_world_the_world_cabi` is defined multiple times
  --> the_world.rs:95:9
   |
81 | macro_rules! __export_world_the_world_cabi{
   | ------------------------------------------ previous definition of the macro `__export_world_the_world_cabi` here
...
95 | pub use __export_world_the_world_cabi;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `__export_world_the_world_cabi` reimported here
   |
   = note: `__export_world_the_world_cabi` must be defined only once in the macro namespace of this module
```

\* *The other errors can be fixed by manually specifying those dependencies, which I omitted for the simplicity of the ticket. For the record, I'm trying to implement [Bazel rules for Wasm modules](https://github.com/actio-cloud/rules_wasm), hence needing to use the CLI rather than crates / Cargo commands.*

You can see the whole generated Rust file [here](https://gist.github.com/ouillie/c62199c77fc6082a941dcf188e9b9b75).

This is where my relatively modest Rust experience holds me back from coming up with a great idea to address the problem. There seems to be no combination of command line arguments to `wit-bindgen` that make it generate a public `export!` macro but also generate Rust code that can compile (due to this "defined multiple times" error). Any pointers would be much appreciated.

**Comment by alexcrichton** [comment]
Thanks for the report! This is a bit of an unfortunate situation since I wish we could generate better code here but long story short the fix is to move `the_world.rs` to a submodule of the crate. If it's the crate root it won't work. For example:

```
$ wit-bindgen rust foo.wit --generate-all --pub-export-macro --default-bindings-module the_world --disable-custom-section-link-helpers
$ echo "mod the_world; pub use the_world::*;" >> my_crate.rs
$ rustc my_crate.rs --crate-type=rlib --edition 2021
```

Note that `--disable-custom-section-link-helpers` disables the runtime dependency on `wit-bindgen-rt` for this smaller example and `--edition 2021` is use to resolve the references to `core`. Does that work for you?

**Comment by ouillie** [comment]
Gotcha, that does work for me! Seems like this was just a misunderstanding of how the output generated from the CLI is meant to be consumed.

Anyway, thanks to your help I now have a first draft of a [`rules_wasm`](https://github.com/actio-cloud/rules_wasm) that should hopefully make it way easier to build components in Bazel repositories.

**Comment by alexcrichton** [comment]
Ah no worries this is no fault of your own, it has to do with how exporting macros has a bit of a history in Rust and in this case unfortunately doesn't work as expected.

Regardless I'm glad it's now working for you! I'm going to close this in the meantime, but feel free to follow up with questions either here or in a new issue.
